flax/core/partial_eval.py

```diff
@@ -0,0 +1,58 @@
+from typing import Any
+
+import functools
+
+import jax
+from jax import linear_util as lu
+from jax.interpreters import partial_eval as pe
+
+from flax import errors
+
+
+def _maybe_unknown(x: Any) -> pe.PartialVal:
+  if isinstance(x, jax.ShapeDtypeStruct):
+    return pe.PartialVal.unknown(jax.ShapedArray(x.shape, x.dtype))
+  else:
+    return pe.PartialVal.known(x)
+
+
+def lazy_init(fn):
+  """Lazily evaluate a function by using the shapes of the inputs.
+
+  The returned function accepts a combination of JAX values and
+  ``jax.ShapeDtypeStruct`` instances for the inputs for which we
+  don't need concrete values (only the shape and dtype).
+
+  This API is used by ``core.lazy_init`` or ``Module.lazy_init``
+  to initialize variables without doing any actual compute on the
```

Comment:
What happens if there exists a Module that initializes variables dependent on some input, and we provide a concrete JAX array for this input when calling lazy init? In that case I think we *do* execute a computation on the inputs (otherwise we cannot create the variables). So maybe it is more correct to say that we don't do any compute on the inputs if the variables do not depend on them?

Reply:
I refer to inputs as the things provided with a ShapeDtypeStruct so doing any computations on them would cause an error (because the value isn't actually available)

Reply:
Why we would allow non-abstract inputs then in the first place? Why not only always ShapeDtypeStructs?